### PR TITLE
Migrate Titan mailbox set up form to new mailbox list component

### PR DESCRIPTION
**Pull request by olaseni**
#### Overview 

The next set of changes 

#### Proposed Changes

* Migrates instances of the provider specific `<TitanNewMailboxList/>` to the generic `<NewMailboxList/>` for the Titan mailbox setup form. This is the form used for creating already purchased mailboxes for users; there's no need for checkout.
* For the main view involved, the JSX components are converted to their TSX counterparts
* Refactors the `useCreateTitanMailboxMutation()` hook such that the internal mutation now accepts variables in the late stage using the `mutateAsync()` function
* Installs autoFocus feature

#### Testing Instructions

* Have a Professional Email subscription with multiple mailboxes ready
* Checkout this branch in your local development setup, or use [Calypso-live ](https://calypso.live/?image=registry.a8c.com/calypso/app:build-35778)
* Navigate to the Email plan page for the subscription i.e. [ **Upgrades** ] > [ **Domains** ] > [ **Active domain** (i.e. example.com) ]  > [ **(\d+) mailboxes** ]
* Open another browser tab to the email provider's external control panel. You can reach that from the Email plan page, using the [ **Manage all mailboxes** ] nav item. Use any of the external nav items exposed to access the external control panel
* Remove one or more mailboxes from the subscription using the external control panel
* Back at the open tab where the Email plan page is open, confirm that the deletion change is reflected ( i.e. one less mailboxes are now displayed)
* Confirm that an `Action required` notice with a CTA to `Set up mailbox` is now visible
*  Use the CTA to reveal the new mailbox form
* Confirm that form's behaviour is identical to the former behaviour before checkout. You can establish this by opening up another tab with the same form using the production website i.e. `https://wordpress.com/***`
* Confirm that you can create a new mailbox for the subscription without going through checkout

| Illustration             |
:------------------:
Screenshot 2022-06-18 at 4 25 09 AM |
Screenshot 2022-06-18 at 4 25 30 AM |
Screenshot 2022-06-18 at 4 26 17 AM |
Screenshot 2022-06-18 at 4 26 43 AM |
Screenshot 2022-06-18 at 4 38 18 AM |

Related to #64424

### CODE REVIEW

File: client/my-sites/email/titan-set-up-mailbox/titan-set-up-mailbox-form.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+const goToThankYouPage = ( siteSlug: string, selectedDomainName: string, mailboxName: string ) => {
+	const emailAddress = `${ mailboxName }@${ selectedDomainName }`;
+	page( emailManagementTitanSetUpThankYou( siteSlug, selectedDomainName, emailAddress ) );
+};
+
+const dispatchCompleteSetupClick = ( dispatchProps: DispatchCompleteSetupClickProps ) => {
+	const { canContinue, dispatch, domainName, mailboxName } = dispatchProps;
+
+	dispatch(
+		recordTracksEvent(
```
Comment (fredrikekelund) [inline client/my-sites/email/titan-set-up-mailbox/titan-set-up-mailbox-form.tsx:41]: I'm curious, are there any rules of thumbs for choosing between `recordTracksEvent` from `calypso/state/analytics/actions` (Redux action creator) vs `recordTracksEvent` from `calypso/lib/analytics/tracks` (push Tracks event "directly")?

Comment (olaseni) [reply client/my-sites/email/titan-set-up-mailbox/titan-set-up-mailbox-form.tsx:41]: None that I know of, personally.

**Comment by fredrikekelund** [comment]
I spent some time looking at the blurring issue I described in [this comment](https://github.com/Automattic/wp-calypso/pull/64605#pullrequestreview-1007716983).

I see that `autoFocus` works as expected with this PR, but I still noticed a temporary de-focusing of the active input field, which bothered me a little bit 😅 It seems like React is re-rendering the entire `NewMailBoxList` subtree fairly frequently and this is what causes the blurring problem. Specifically, the `useGetExistingMailboxNames` hook seems to be the culprit. Likely, the problem has got something to do with the internal `isLoading` etc flags in react-query. 

Here's a screencast that demonstrates how the entire subtree of `NewMailBoxList` re-renders excessively (notice how the Chrome devtools highlight those DOM nodes as I just move in/out of the viewport with the cursor).

https://user-images.githubusercontent.com/1101677/174629710-16c70399-0f16-488f-afc2-efe8e8c99712.mov

I have a suggestion that seems to fix the problem, which is to change the arguments passed to `useGetEmailAccountsQuery` in `client/my-sites/email/form/mailboxes/components/new-mailbox-list/use-get-existing-mailbox-names.ts`:

```ts
useGetEmailAccountsQuery( selectedSiteId as number, selectedDomainName, {
    notifyOnChangeProps: [ 'data' ],
} )
```

I can't see why the internal flags of the react-query hook would cause the entire DOM tree to re-render, so I'm not sure that this is a 100% satisfying solution, but still it gets to the issue at hand.

File: client/my-sites/email/titan-set-up-mailbox/titan-set-up-mailbox-form.tsx (hunk 1 above)
Comment (fredrikekelund) [reply client/my-sites/email/titan-set-up-mailbox/titan-set-up-mailbox-form.tsx:41]: Okay, cool! Wasn't necessarily a comment to change this code. I've seen plenty of both solutions in the codebase and I had to ask just because it seems simpler to not have to `dispatch` the event.

Comment (stephanethomas) [reply client/my-sites/email/titan-set-up-mailbox/titan-set-up-mailbox-form.tsx:41]: Let's ping @Automattic/team-calypso, maybe they can clarify which option is better 🙂.

File: client/data/emails/use-create-titan-mailbox-mutation.ts [hunk 2]
```diff
@@ -1,41 +1,45 @@
...
-			} ),
-		{ ...mutationOptions, mutationKey: `${ domainName }@${ mailboxName }` }
+			} );
+		}
 	);
-}
+};
+
+export { useCreateTitanMailboxMutation };
+export type { TitanMailboxFields };
```
Comment (stephanethomas) [inline client/data/emails/use-create-titan-mailbox-mutation.ts]: We could remove those two lines by exporting the type and function directly, e.g.:

```javascript
export const useCreateTitanMailboxMutation = () => { ... }
```

It also makes it clear which element is exported looking at their declaration.

File: client/my-sites/email/form/mailboxes/components/mailbox-field/index.tsx [hunk 3]
```diff
@@ -46,6 +48,7 @@ const MailboxFieldInput = ( {
 		onBlur,
 		onChange,
 		onInvalid,
+		...( renderPosition === 1 ? { autoFocus: true } : {} ),
```
Comment (stephanethomas) [inline client/my-sites/email/form/mailboxes/components/mailbox-field/index.tsx]: A small suggestion but instead of passing `renderPosition` we could actually provide a boolean called `isFirstField`, which is a bit more explicit than the former.

File: client/data/emails/use-create-titan-mailbox-mutation.ts [hunk 4]
```diff
@@ -1,41 +1,45 @@
...
+} from 'calypso/my-sites/email/form/mailboxes/constants';
+
+type TitanMailboxFields = {
+	[ FIELD_DOMAIN ]: string;
+	[ FIELD_NAME ]: string;
+	[ FIELD_MAILBOX ]: string;
+	[ FIELD_PASSWORD ]: string;
+	[ FIELD_ALTERNATIVE_EMAIL ]: string;
+	[ FIELD_IS_ADMIN ]: boolean;
+};
```
Comment (stephanethomas) [inline client/data/emails/use-create-titan-mailbox-mutation.ts]: We could move that type declaration to the `types.ts` file (in the same folder, or even better, in `calypso/my-sites/email/form/mailboxes/constants`).

File: client/my-sites/email/form/mailboxes/components/mailbox-form-wrapper/index.tsx [hunk 5]
```diff
@@ -27,15 +27,20 @@ const MailboxFormWrapper = ( {
...
 		return {
 			field,
 			onFieldValueChanged,
 			onRequestFieldValidation: () => mailbox.validateField( field.fieldName ),
+			renderPosition,
 		};
 	};
 
-	const UserFormFields = () => {
+	const renderUserFormFields = () => {
```
Comment (stephanethomas) [inline client/my-sites/email/form/mailboxes/components/mailbox-form-wrapper/index.tsx:41]: What's the reason for renaming this? It's no longer consistent with how we named the `GoogleFormFields` and `TitanFormFields` components.

File: client/my-sites/email/titan-set-up-mailbox/index.tsx [hunk 6]
```diff
@@ -56,10 +65,10 @@ const TitanSetUpMailbox = ( { selectedDomainName, source } ) => {
 
 	return (
 		<>
-			{ selectedSite && <QuerySiteDomains siteId={ selectedSite.ID } /> }
+			{ selectedSiteId && <QuerySiteDomains siteId={ selectedSiteId } /> }
 
 			<Main wideLayout={ true }>
-				<DocumentHead title={ titleCase( title ) } />
+				<DocumentHead title={ title } />
```
Comment (stephanethomas) [inline client/my-sites/email/titan-set-up-mailbox/index.tsx]: Why did you remove `titleCase()`? It was here on purpose 😉.

**Review by stephanethomas — APPROVED** [review]
I only had a few minor comments but otherwise this looks good, and works fine 👍.

image

File: client/my-sites/email/form/mailboxes/components/mailbox-form-wrapper/index.tsx (hunk 5 above)
Comment (olaseni) [reply client/my-sites/email/form/mailboxes/components/mailbox-form-wrapper/index.tsx:41]: When it became important to dynamically determine the position of the form elements, I realised that inner components are essentially functions and are executed after the currently executing component has been fully evaluated.

```jsx
<>
    <div> Words and text </div>
    <DirectComponent />
    { someFunctionThatRendersComponents() }
</>
```
For example, `<DirectComponent />` will be rendered/called in its own context AFTER the `someFunctionThatRendersComponents()` is called. 

This makes determining visual element positions difficult. 
The workaround is to call a `renderer function` because then the scope of that function evaluates in the right order.

File: client/my-sites/email/titan-set-up-mailbox/index.tsx (hunk 6 above)
Comment (olaseni) [reply client/my-sites/email/titan-set-up-mailbox/index.tsx]: There's no type definition for that module since it's an older one and seemingly hasn't been used in TS before. 
I figured we could achieve the same in CSS since it's such a trivial case.

However I have reconsidered and added a local type definition file that declares that module as `any`. I hope that's okay @Automattic/team-calypso ?

**Comment by olaseni** [comment]
> I spent some time looking at the blurring issue I described in [this comment](https://github.com/Automattic/wp-calypso/pull/64605#pullrequestreview-1007716983).
 ...  I can't see why the internal flags of the react-query hook would cause the entire DOM tree to re-render, so I'm not sure that this is a 100% satisfying solution, but still it gets to the issue at hand.

Thanks for looking into this. @fredrikekelund . I suspected this as well but was still working on a good fix. I'll add this in since it's better and works now.

File: client/my-sites/email/titan-set-up-mailbox/titan-set-up-mailbox-form.tsx (hunk 1 above)
Comment (tyxla) [reply client/my-sites/email/titan-set-up-mailbox/titan-set-up-mailbox-form.tsx:41]: Hey folks, thanks for the ping! 

We'd generally recommend **going with `lib/analytics` when possible**, because `state/analytics` is simply a wrapper around it, and we'd generally like to get rid of any unnecessary wrappers.

The rationale is simple: tracks code is spread throughout the codebase, and if we don't rely on Redux and all the underlying custom infrastructure, we have one less piece to carry around throughout the codebase.

Now, what likely makes this confusing is that there was a time when we did recommend using the Redux wrapper. However, that was in times when we worked to centralize Redux and the Calypso implementation as "one solution to rule them all". With `react-query ` and `@wordpress/data` this is no longer the case, and the less coupled we are, the better.

File: client/my-sites/email/titan-set-up-mailbox/index.tsx (hunk 6 above)
Comment (tyxla) [reply client/my-sites/email/titan-set-up-mailbox/index.tsx]: FWIW we can't style the title of the document with CSS - it's fixed by the browser, and specifically, the OS we're using.  

When it comes to declaring a local module - that generally works, but can we make it a bit more specific than just `any`? It seems to be a straightforward case that just accepts strings and returns strings?

Comment (olaseni) [reply client/my-sites/email/titan-set-up-mailbox/index.tsx]: > FWIW we can't style the title of the document with CSS - it's fixed by the browser, and specifically, the OS we're using

Oh, sorry about the confusion. My reference to styling was to the `<SectionHeader/>`

Comment (olaseni) [reply client/my-sites/email/titan-set-up-mailbox/index.tsx]: > but can we make it a bit more specific than just any? It seems to be a straightforward case that just accepts strings and returns strings?

Sure. Thanks for the quick feedback!
